### ISSUE TITLE
assume_timezone can only handle IANA timezone names, not fixed offsets

### ISSUE THREAD

**Issue by sm-Fifteen** [apache/arrow#36558]
### Describe the bug, including details regarding any error messages, version, and platform.

See https://github.com/pola-rs/polars/issues/9586#issuecomment-1625418583, where the issue was identified.

```py
import pyarrow as pa
import pyarrow.compute as pc

from datetime import datetime

pc.assume_timezone(pa.array([datetime(2020, 1, 1)]), '+00:00')
```

> ArrowInvalid: Cannot locate timezone '+00:00': +00:00 not found in timezone database

The [Arrow format specification](https://github.com/apache/arrow/blob/dd3670583d6fd8b95783f89f7b80c04588e16fb1/format/Schema.fbs#L319-L361) describes 3 timestamp formats: "naive date-time" (timezone string is null), "zoned date-time" (timezone string is from tzdb) and "offset date-time" (timezone string is a fixed RFC 3339 num-offset, so no Z). [The doc for assume_timezone](https://arrow.apache.org/docs/python/generated/pyarrow.compute.assume_timezone.html) makes no specific mention of it, but it cannot handle being passed an offset because it [only performs a lookup in tzdb](https://github.com/apache/arrow/blob/f8256bd625ac0b06238011fc13ea0249956e3859/cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc#L116-L134) (via `LocateZone`) and doesn't try to parse the offset. That's despite documentation in other places saying that [offsets are completely fine and that "+00:00" should be considered as identical to "UTC"](https://arrow.apache.org/docs/cpp/api/datatype.html#classarrow_1_1_timestamp_type). Given all this, I would expect "+00:00" to be properly recognized as UTC, and "+01:00" to be recognized as a fixed offset.

### Component(s)

Python

**Comment by westonpace** [comment]
This may be a duplicate of #30036 (though I feel like this ticket is worded more clearly).

**Comment by buhrmann** [comment]
Related to this (in the sense that offsets are inconsistently supported in Arrow), one can also cast with offsets that are later invalid, e.g. when converting to pandas. For example, while this works:

``` python
arr = pa.array([datetime(2021, 1, 1)]).cast(pa.timestamp(unit="ns", tz="+04:00"))
print(arr.to_pandas())
```

This doesn't:
``` python
arr = pa.array([datetime(2021, 1, 1)]).cast(pa.timestamp(unit="ns", tz="+0400"))
print(arr.to_pandas())
```
It wil happily cast with offset +0400 but will then fail in the conversion to pandas:
```
File [~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/array.pxi:867](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/array.pxi:867), in pyarrow.lib._PandasConvertible.to_pandas()

File [~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/array.pxi:1491](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/array.pxi:1491), in pyarrow.lib.Array._to_pandas()

File [~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/array.pxi:1741](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/array.pxi:1741), in pyarrow.lib._array_like_to_pandas()

File [~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1215](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1215), in make_tz_aware(series, tz)
   [1211](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1211) def make_tz_aware(series, tz):
   [1212](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1212)     """
   [1213](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1213)     Make a datetime64 Series timezone-aware for the given tz
   [1214](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1214)     """
-> [1215](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1215)     tz = pa.lib.string_to_tzinfo(tz)
   [1216](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1216)     series = (series.dt.tz_localize('utc')
   [1217](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1217)                     .dt.tz_convert(tz))
   [1218](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1218)     return series

File [~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/types.pxi:3576](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/types.pxi:3576), in pyarrow.lib.string_to_tzinfo()

File [~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/error.pxi:144](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pyarrow/error.pxi:144), in pyarrow.lib.pyarrow_internal_check_status()

File [~/micromamba/envs/grp/lib/python3.9/site-packages/pytz/__init__.py:188](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pytz/__init__.py:188), in timezone(zone)
    [186](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pytz/__init__.py:186)             fp.close()
    [187](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pytz/__init__.py:187)     else:
--> [188](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pytz/__init__.py:188)         raise UnknownTimeZoneError(zone)
    [190](https://file+.vscode-resource.vscode-cdn.net/Users/thomas/code/notebooks/~/micromamba/envs/grp/lib/python3.9/site-packages/pytz/__init__.py:190) return _tzinfo_cache[zone]

UnknownTimeZoneError: '+0400'
```

Note that +0400 is a valid ISO offset, and pyarrow will itself create timestamp types with this as the `tz` parameter when reading a CSV, for example.

**Comment by MarcoGorelli** [comment]
To be honest I'm hoping Arrow won't support fixed offsets

For individual elements, sure, they're well defined and unambiguous. `'2020-01-01T01:12:34+05:00'` is unambiguous

But if you're working with timezone-aware data, you're presumably working with a real-world problem - in which case, nudging you towards using the `Area/Location` time zone identifier will prevent you from ending up with daylight-saving-time-related issues if you'd mistakenly used a fixed offset

For example, '2023-02-16T17:19:02+02:00' could be both 'Europe/Sofia' (which observes daylight saving time) or 'Africa/Johannesburg' (which doesn't observe daylight saving time). The individual timestamp is allowed to have `+02:00`, but letting the dtype of the array be a fixed offset sounds like a user bug waiting to happen

**Comment by sm-Fifteen** [comment]
@MarcoGorelli: What would be the difference between setting a column to have `tz='+02:00'` compared to the completely valid `tz='Etc/GMT+2'` in this context? Both would have the same effect, it's just that IANA doesn't have zones for sub-hour precision offsets in the few cases where that's needed. I get that it's preferable that users know the reference timezone they're working with for things like DST policy reasons, and that the internal representation will be UTC regardless of what the TZ specifier is, but restricting it like this (for admittedly fair "footgun" reasons) seems like the wrong solution, here.

**Comment by MarcoGorelli** [comment]
> it's just that IANA doesn't have zones for sub-hour precision offsets in the few cases where that's needed

could you clarify when this is needed please?

**Comment by sm-Fifteen** [comment]
> > it's just that IANA doesn't have zones for sub-hour precision offsets in the few cases where that's needed
> 
> could you clarify when this is needed please?

According to [Time and Date](https://www.timeanddate.com/time/time-zones-interesting.html), Australia/Adelaide, Asia/Calcutta (India) and America/St_Johns (Newfoundland, Canada), among others, have half-hour offsets from UTC, and their normal time zone reflects that, but they can't be represented at whole hour offsets with Etc zones from tzdb.

**Comment by buhrmann** [comment]
The other thing is that Arrow itself doesn't support timezone identifiers in the Area/Location format anywhere that data is ingested. Neither CSVs, for example, nor the cast support it. That's a whole other issue, you can only get timestamps into Arrow using timezone offsets, but then they're useless, because everywhere you can actually use a timezone it only works with area identifiers.

**Comment by sm-Fifteen** [comment]
> The other thing is that Arrow itself doesn't support timezone identifiers in the Area/Location format anywhere that data is ingested. Neither CSVs, for example, nor the cast support it. That's a whole other issue, you can only get timestamps into Arrow using timezone offsets, but then they're useless, because everywhere you can actually use a timezone it only works with area identifiers.

Yeah, unless you store the timezone offset into a separate column (I don't know whether that's possible when using Arrow directly), when loading timestamps from a row-oriented format like CSV, you lose that information when the entire column is normalized. To get a timezone specifier, assigned to the column, you have to know it in advance or disregard the original one and use a "working/display timezone".

Even if you could load timestamps from a CSV file with IANA timezones attached to them, like with the [proposed ECMAScript extended timestamp format](https://tc39.es/proposal-temporal/docs/persistence-model.svg), you would have the same problem as with plain offsets, because the issue isn't using or not using tzdb names, it's whether or not the column can be considered to use a single, uniform timezone.

**Comment by MarcoGorelli** [comment]
> it's whether or not the column can be considered to use a single, uniform timezone

I don't think Arrow (nor any dataframe library, as far as I'm aware) supports columns with mixed timezones (barring `object` dtype in pandas, which can store anything, but then time series algorithms don't work on it)

**Comment by buhrmann** [comment]
Yes, I'm currently working on recovering the original time zone after ingesting data, if and only if the initial string representations of the timestamps contained a single unique offset. It works reliably enough, but since I can only parse timestamps with fixed offsets in Arrow, the offsets are also the only `tz` representation I can recover. But I understand your point that identifiers would be more useful, and that even then there would be no way currently to represent mixed timezones without extracting the timezone into a separate column.

**Comment by sm-Fifteen** [comment]
> > 
> 
> I don't think Arrow (nor any dataframe library, as far as I'm aware) supports columns with mixed timezones (barring `object` dtype in pandas, which can store anything, but then time series algorithms don't work on it)

@MarcoGorelli : Oh, obvously working with mixed timezone or mixed offset series is going to prove pretty much unworkable for most uses if you don't normalize it first, and I'm also pretty sure Arrow doesn't have a column format that could handle mixed zone timestamps. I'm just saying that It's not a problem that's exclusive to fixed time offsets, that the problem can pose itself even if you're working with IANA zones (such as a vehicle moving across timezones while measuring something, and the logs adjusting the zone consequently) and that it's not a good reason for Arrow to not support offset timezones.

If you're dealing with a column that has a uniform, fixed offset time, that shouldn't be forbidden. Mixed offsets and/or mixed timezones are, I believe, a separate problem. In fact, I would go as far as to suggest Arrow might want to add a fourth timestamp format, which would see the internal representation normalized to UTC (as with the other two tz-aware formats) and the tz string in the column metadata pointing to a separate column containing the timezone names or fixed offsets.

**Comment by rok** [comment]
> [..] I would go as far as to suggest Arrow might want to add a fourth timestamp format, which would see the internal representation normalized to UTC (as with the other two tz-aware formats) and the tz string in the column metadata pointing to a separate column containing the timezone names or fixed offsets.

This would could be done with a (cannonical) extension type as it is a bit niche.

That said #12865 should address this issue.

**Comment by CurtHagenlocher** [comment]
> If you're dealing with a column that has a uniform, fixed offset time, that shouldn't be forbidden. Mixed offsets and/or mixed timezones are, I believe, a separate problem. In fact, I would go as far as to suggest Arrow might want to add a fourth timestamp format, which would see the internal representation normalized to UTC (as with the other two tz-aware formats) and the tz string in the column metadata pointing to a separate column containing the timezone names or fixed offsets.

I think it would make more sense to do this as a struct where one field has the time normalized to UTC and the other has the time zone and/or offset.

This is related to #44248.

**Comment by rok** [comment]
> I think it would make more sense to do this as a struct where one field has the time normalized to UTC and the other has the time zone and/or offset.
> 
> This is related to [#44248](https://github.com/apache/arrow/issues/44248).

This (per-row offset) proposal was since voted and merge and this issue was resolved with https://github.com/apache/arrow/pull/12865 too, so the following behavior is now in place:

```python
>>> import pyarrow as pa
>>> import pyarrow as pa
>>> import pyarrow.compute as pc
>>>
>>> from datetime import datetime
>>>
>>> pc.assume_timezone(pa.array([datetime(2020, 1, 1)]), '+12:34')
<pyarrow.lib.TimestampArray object at 0x7ba8e0287460>
[
  2019-12-31 11:26:00.000000Z
]
```
